File: src/parser_recibos.py
"""Parser for Marval receipt PDFs (recibo_contrib_v4.pdf, recibo_contrib_v4_rrhh.pdf)."""
import re
import pdfplumber
from typing import Dict, List, Optional

from .models import Concepto, ReciboEmpleado

MESES = (
    r'(?:Enero|Febrero|Marzo|Abril|Mayo|Junio|'
    r'Julio|Agosto|Septiembre|Octubre|Noviembre|Diciembre)'
)


def parse_money(s: str) -> Optional[float]:
    """Convert Argentine or US money string to float.

    Handles: '$ 1.234.567,89' (AR) and '$10,963,803.18' (US).
    """
    if not s:
        return None
    s = re.sub(r'[$\s]', '', str(s)).strip()
    if not s or s in ('-', ''):
        return None
    # US format: ends with .XX (two decimal digits after dot)
    if re.match(r'^-?[\d,]+\.\d{1,2}$', s):
        return float(s.replace(',', ''))
    # AR format: ends with ,XX
    if re.match(r'^-?[\d.]+,\d{1,2}$', s):
        return float(s.replace('.', '').replace(',', '.'))
    # Plain integer
    if re.match(r'^-?\d+$', s):
        return float(s)
    return None


def _parse_concepto_line(line: str) -> Optional[Concepto]:
    """Parse 'CODE Description [UNIT] $ AMOUNT' -> Concepto."""
    m = re.match(r'^(-?\d{3,6})\s+(.+?)\s+\$\s*(-?[\d.,]+)\s*$', line.strip())
    if not m:
        return None
    code = m.group(1)
    raw_desc = m.group(2).strip()
    amount_str = m.group(3)
    # Strip trailing unit/base numbers from description (e.g. "Jubilación 11,00")
    desc = re.sub(r'\s+\d{1,3}(?:,\d+)?\s*$', '', raw_desc).strip()
    amount = parse_money(amount_str)
    if amount is None:
        return None
    return Concepto(codigo=code, descripcion=desc, monto=amount)


def _parse_page(text: str, page_num: int) -> Optional[ReciboEmpleado]:
    lines = [ln.strip() for ln in text.split('\n')]

    rp = ReciboEmpleado(
        legajo='', nombre='',
        bruto=None, neto=None,
        total_contribuciones=None, costo_empleador=None,
        composicion_rem=None, composicion_no_rem=None, composicion_desc=None,
        paginas=[page_num],
    )

    # State machine
    state = 'HEADER'
    concepto_section_count = 0  # track which CONCEPTO block we're in

    for line in lines:
        if not line:
            continue

        # --- HEADER: find legajo + nombre + bruto ---
        if state == 'HEADER':
            m = re.search(
                rf'{MESES}\s+\d{{4}}\s+(.+?)\s+(\d{{3,6}})\s+\$\s*([\d.,]+)',
                line,
            )
            if m:
                rp.nombre = m.group(1).strip()
                # Normalizar legajo igual que la liquidación (sin ceros a la
                # izquierda) para que '0826' (recibo) matchee '826' (liqui).
                rp.legajo = (m.group(2).strip().lstrip('0') or '0')
                state = 'PRE_CONTRIB'
                continue

        # --- COSTO TOTAL EMPLEADOR (can appear anywhere before contributions) ---
        if state in ('HEADER', 'PRE_CONTRIB', 'CONTRIB'):
            m = re.search(r'COSTO TOTAL EMPLEADOR\s+\$\s*([\d.,]+)', line)
            if m:
                rp.costo_empleador = parse_money(m.group(1))

        # --- Start of first CONCEPTO block = contributions section ---
        if state == 'PRE_CONTRIB' and line == 'CONCEPTO UNIDAD BASE MONTO':
            state = 'CONTRIB'
            continue

        if state == 'CONTRIB':
            m = re.search(r'SUB TOTAL CONTRIBUCIONES EMPLEADOR\s+\$\s*([\d.,]+)', line)
            if m:
                rp.total_contribuciones = parse_money(m.group(1))
                state = 'PRE_CONCEPTOS'
                continue
            c = _parse_concepto_line(line)
            if c:
                rp.contribuciones.append(c)

        # --- Between contributions and concepts: get SUELDO BRUTO ---
        if state == 'PRE_CONCEPTOS':
            m = re.match(r'SUELDO BRUTO\s+\$\s*([\d.,]+)', line)
            if m:
                rp.bruto = parse_money(m.group(1))
                continue
            if line == 'CONCEPTO UNIDAD BASE MONTO':
                state = 'CONCEPTOS'
                continue

        # --- Haberes / Descuentos section ---
        if state == 'CONCEPTOS':
            # COMPOSICION SALARIAL marks end of concepts
            m = re.search(
                r'Remunerativo:\s*\$\s*([\d,\.]+)'
                r'\s+No Remunerativo:\s*\$\s*([\d,.]+)'
                r'\s+Descuentos:\s*\$\s*([\d,.]+)',
                line,
            )
            if m:
                rp.composicion_rem = parse_money(m.group(1))
                rp.composicion_no_rem = parse_money(m.group(2))
                rp.composicion_desc = parse_money(m.group(3))
                state = 'POST_CONCEPTOS'
                continue
            c = _parse_concepto_line(line)
            if c:
                rp.conceptos.append(c)

        # --- After COMPOSICION SALARIAL: find SUELDO NETO ---
        if state == 'POST_CONCEPTOS':
            m = re.match(r'SUELDO NETO\s+\$\s*([\d.,]+)', line)
            if m:
                rp.neto = parse_money(m.group(1))
                state = 'PIE'
                continue

        # --- Pie chart percentages ---
        if state == 'PIE':
            for pct in re.findall(r'(\d{1,2}\.\d{2})%', line):
                rp.porcentajes_torta.append(float(pct))

    if not rp.legajo:
        rp.errores_parse.append(f'Página {page_num}: no se detectó legajo')
        return None

    return rp


def _merge_pages(base: ReciboEmpleado, extra: ReciboEmpleado) -> None:
    """Add extra page data into base (for multi-page employees)."""
    base.conceptos.extend(extra.conceptos)
    base.contribuciones.extend(extra.contribuciones)
    base.paginas.extend(extra.paginas)
    base.n_paginas += 1
    base.porcentajes_torta.extend(extra.porcentajes_torta)

    for field in ('bruto', 'neto', 'total_contribuciones', 'costo_empleador'):
        bv = getattr(base, field)
        ev = getattr(extra, field)
        if bv is not None and ev is not None:
            setattr(base, field, round(bv + ev, 2))
        elif ev is not None:
            setattr(base, field, ev)

    for field in ('composicion_rem', 'composicion_no_rem', 'composicion_desc'):
        bv = getattr(base, field)
        ev = getattr(extra, field)
        if bv is not None and ev is not None:
            setattr(base, field, round(bv + ev, 2))
        elif ev is not None:
            setattr(base, field, ev)


def parse_recibos(pdf_paths: List[str], verbose: bool = False) -> Dict[str, ReciboEmpleado]:
    """Parse one or more receipt PDFs. Returns dict keyed by legajo string."""
    results: Dict[str, ReciboEmpleado] = {}
    sin_legajo = 0

    for pdf_path in pdf_paths:
        with pdfplumber.open(pdf_path) as pdf:
            for page_num, page in enumerate(pdf.pages, 1):
                text = page.extract_text() or ''
                rp = _parse_page(text, page_num)
                if rp is None:
                    sin_legajo += 1
                    if verbose:
                        print(f'[WARN] {pdf_path} pág {page_num}: sin legajo detectado')
                    continue
                if rp.legajo in results:
                    _merge_pages(results[rp.legajo], rp)
                    if verbose:
                        print(f'[INFO] Legajo {rp.legajo} (pág {page_num}): bloque adicional sumado')
                else:
                    results[rp.legajo] = rp

    if verbose:
        print(f'[INFO] Recibos parseados: {len(results)} empleados, {sin_legajo} páginas sin legajo')

    return results

File: docs/parsers/parser-recibos.js
// parser-recibos.js
// Parser de PDFs de recibos Marval (recibo_contrib_v4.pdf, recibo_contrib_v4_rrhh.pdf).
// Port fiel desde src/parser_recibos.py.
//
// Módulo ES, sin dependencias de DOM/window: importable en Node (ESM) y en el
// navegador (<script type="module">). La extracción de texto del PDF se hace
// aparte (ver pdf-extract.js); aquí el caller inyecta el texto ya extraído.

// Alternativa de mes (igual que MESES en el Python). Sin grupo de captura: usa (?:...).
const MESES =
  '(?:Enero|Febrero|Marzo|Abril|Mayo|Junio|' +
  'Julio|Agosto|Septiembre|Octubre|Noviembre|Diciembre)';

// Redondeo a 2 decimales (equivalente a round(x, 2) del Python para los montos
// que maneja este parser). Nota: Python usa banker's rounding; JS Math.round
// redondea ".5" hacia arriba. Para los importes contables involucrados la
// diferencia no se observa; cualquier divergencia se verifica por separado.
function _round2(x) {
  return Math.round((x + Number.EPSILON) * 100) / 100;
}

/**
 * Convierte un string de dinero argentino o US a number.
 *
 * Maneja: '$ 1.234.567,89' (AR) y '$10,963,803.18' (US).
 * Devuelve null cuando el Python devuelve None.
 */
export function parseMoney(s) {
  if (!s) {
    return null;
  }
  // re.sub(r'[$\s]', '', str(s)).strip() -> elimina '$' y todo whitespace.
  s = String(s).replace(/[$\s]/g, '').trim();
  if (!s || s === '-' || s === '') {
    return null;
  }
  // Formato US: termina en .XX (uno o dos dígitos decimales tras el punto).
  if (/^-?[\d,]+\.\d{1,2}$/.test(s)) {
    return parseFloat(s.replace(/,/g, ''));
  }
  // Formato AR: termina en ,XX.
  if (/^-?[\d.]+,\d{1,2}$/.test(s)) {
    return parseFloat(s.replace(/\./g, '').replace(',', '.'));
  }
  // Entero plano.
  if (/^-?\d+$/.test(s)) {
    return parseFloat(s);
  }
  return null;
}

/**
 * Parsea 'CODE Description [UNIT] $ AMOUNT' -> Concepto, o null.
 * @returns {{codigo:string, descripcion:string, monto:number, columna:string}|null}
 */
function _parseConceptoLine(line) {
  // re.match -> anclado al inicio. El patrón ya incluía '^' y '$'.
  const m = line.trim().match(/^(-?\d{3,6})\s+(.+?)\s+\$\s*(-?[\d.,]+)\s*$/);
  if (!m) {
    return null;
  }
  const code = m[1];
  const rawDesc = m[2].trim();
  const amountStr = m[3];
  // Quita números de unidad/base al final de la descripción (ej. "Jubilación 11,00").
  const desc = rawDesc.replace(/\s+\d{1,3}(?:,\d+)?\s*$/, '').trim();
  const amount = parseMoney(amountStr);
  if (amount === null) {
    return null;
  }
  return { codigo: code, descripcion: desc, monto: amount, columna: '' };
}

// Crea un ReciboEmpleado con los defaults del dataclass del Python.
function _nuevoRecibo(pageNum) {
  return {
    legajo: '',
    nombre: '',
    bruto: null,
    neto: null,
    total_contribuciones: null,
    costo_empleador: null,
    composicion_rem: null,
    composicion_no_rem: null,
    composicion_desc: null,
    conceptos: [],
    contribuciones: [],
    porcentajes_torta: [],
    paginas: [pageNum],
    n_paginas: 1,
    errores_parse: [],
  };
}

/**
 * Parsea una página (texto) -> ReciboEmpleado o null.
 */
function _parsePage(text, pageNum) {
  const lines = text.split('\n').map((ln) => ln.trim());

  const rp = _nuevoRecibo(pageNum);

  // Máquina de estados.
  let state = 'HEADER';

  for (const line of lines) {
    if (!line) {
      continue;
    }

    // --- HEADER: busca legajo + nombre + bruto ---
    if (state === 'HEADER') {
      const reHeader = new RegExp(
        MESES + '\\s+\\d{4}\\s+(.+?)\\s+(\\d{3,6})\\s+\\$\\s*([\\d.,]+)'
      );
      const m = line.match(reHeader);
      if (m) {
        rp.nombre = m[1].trim();
        // Normalizar legajo igual que la liquidación (sin ceros a la
        // izquierda) para que '0826' (recibo) matchee '826' (liqui).
        rp.legajo = m[2].trim().replace(/^0+/, '') || '0';
        state = 'PRE_CONTRIB';
        continue;
      }
    }

    // --- COSTO TOTAL EMPLEADOR (puede aparecer en cualquier lugar antes de contribuciones) ---
    if (state === 'HEADER' || state === 'PRE_CONTRIB' || state === 'CONTRIB') {
      const m = line.match(/COSTO TOTAL EMPLEADOR\s+\$\s*([\d.,]+)/);
      if (m) {
        rp.costo_empleador = parseMoney(m[1]);
      }
    }

    // --- Inicio del primer bloque CONCEPTO = sección de contribuciones ---
    if (state === 'PRE_CONTRIB' && line === 'CONCEPTO UNIDAD BASE MONTO') {
      state = 'CONTRIB';
      continue;
    }

    if (state === 'CONTRIB') {
      const m = line.match(/SUB TOTAL CONTRIBUCIONES EMPLEADOR\s+\$\s*([\d.,]+)/);
      if (m) {
        rp.total_contribuciones = parseMoney(m[1]);
        state = 'PRE_CONCEPTOS';
        continue;
      }
      const c = _parseConceptoLine(line);
      if (c) {
        rp.contribuciones.push(c);
      }
    }

    // --- Entre contribuciones y conceptos: obtener SUELDO BRUTO ---
    if (state === 'PRE_CONCEPTOS') {
      const m = line.match(/^SUELDO BRUTO\s+\$\s*([\d.,]+)/);
      if (m) {
        rp.bruto = parseMoney(m[1]);
        continue;
      }
      if (line === 'CONCEPTO UNIDAD BASE MONTO') {
        state = 'CONCEPTOS';
        continue;
      }
    }

    // --- Sección Haberes / Descuentos ---
    if (state === 'CONCEPTOS') {
      // COMPOSICION SALARIAL marca el fin de los conceptos.
      const m = line.match(
        /Remunerativo:\s*\$\s*([\d,.]+)\s+No Remunerativo:\s*\$\s*([\d,.]+)\s+Descuentos:\s*\$\s*([\d,.]+)/
      );
      if (m) {
        rp.composicion_rem = parseMoney(m[1]);
        rp.composicion_no_rem = parseMoney(m[2]);
        rp.composicion_desc = parseMoney(m[3]);
        state = 'POST_CONCEPTOS';
        continue;
      }
      const c = _parseConceptoLine(line);
      if (c) {
        rp.conceptos.push(c);
      }
    }

    // --- Después de COMPOSICION SALARIAL: buscar SUELDO NETO ---
    if (state === 'POST_CONCEPTOS') {
      const m = line.match(/^SUELDO NETO\s+\$\s*([\d.,]+)/);
      if (m) {
        rp.neto = parseMoney(m[1]);
        state = 'PIE';
        continue;
      }
    }

    // --- Porcentajes del gráfico de torta ---
    if (state === 'PIE') {
      const matches = line.matchAll(/(\d{1,2}\.\d{2})%/g);
      for (const pct of matches) {
        rp.porcentajes_torta.push(parseFloat(pct[1]));
      }
    }
  }

  if (!rp.legajo) {
    rp.errores_parse.push(`Página ${pageNum}: no se detectó legajo`);
    return null;
  }

  return rp;
}

/**
 * Suma los datos de la página extra dentro de base (empleados multi-página).
 */
function _mergePages(base, extra) {
  base.conceptos.push(...extra.conceptos);
  base.contribuciones.push(...extra.contribuciones);
  base.paginas.push(...extra.paginas);
  base.n_paginas += 1;
  base.porcentajes_torta.push(...extra.porcentajes_torta);

  for (const field of ['bruto', 'neto', 'total_contribuciones', 'costo_empleador']) {
    const bv = base[field];
    const ev = extra[field];
    if (bv !== null && ev !== null) {
      base[field] = _round2(bv + ev);
    } else if (ev !== null) {
      base[field] = ev;
    }
  }

  for (const field of ['composicion_rem', 'composicion_no_rem', 'composicion_desc']) {
    const bv = base[field];
    const ev = extra[field];
    if (bv !== null && ev !== null) {
      base[field] = _round2(bv + ev);
    } else if (ev !== null) {
      base[field] = ev;
    }
  }
}

/**
 * Parsea uno o más PDFs de recibos. Devuelve un objeto-mapa indexado por legajo.
 *
 * @param {string[][]} pagesByFile - un array por archivo PDF, con el texto de
 *   cada página (string) en orden.
 * @returns {Object<string, Object>} mapa { [legajo]: ReciboEmpleado }
 */
export function parseRecibos(pagesByFile) {
  const results = {};

  for (const pages of pagesByFile) {
    // page_num arranca en 1 por cada archivo (igual que enumerate(pdf.pages, 1)).
    let pageNum = 0;
    for (const page of pages) {
      pageNum += 1;
      const text = page || '';
      const rp = _parsePage(text, pageNum);
      if (rp === null) {
        continue;
      }
      if (Object.prototype.hasOwnProperty.call(results, rp.legajo)) {
        _mergePages(results[rp.legajo], rp);
      } else {
        results[rp.legajo] = rp;
      }
    }
  }

  return results;
}
